### DISCUSSION TITLE
uasyncio code - memory usage constantly increasing

### DISCUSSION THREAD

**Discussion by steeley** [micropython/micropython#11488] [Using MicroPython] [answered]
just trying some uasyncio experiments on an ESP32-c3. My code is based on what I can find and understand from examples is show below. It seems to continually use memory more and more as it loops
I made an example not using the get_event_loop stuff ( shown further down) and this is fine memory use is stable.
I need to have the program run forever, so any ideas how this can work correctly welcome.

TIA

```
import uasyncio as asyncio
from machine import Pin
import micropython

pin6 = Pin(6, Pin.OUT)
pin7 = Pin(7, Pin.OUT)

async def p_task():
    while True:
        #print("p_task")
        pin6.value(1)
        await asyncio.sleep(0.25)
        pin6.value(0)
        await asyncio.sleep(0.25)
        
        
async def q_task():
    while True:
        #print("q_task")
        pin7.value(1)
        await asyncio.sleep(0.6)
        pin7.value(0)
        await asyncio.sleep(0.6)
        
async def info():
    while True:
        print(micropython.mem_info())
        await asyncio.sleep(2)
    
loop = asyncio.get_event_loop()
loop.create_task(p_task())
loop.create_task(q_task())
loop.create_task(info())

loop.run_forever()
```

This example is ok and does not use up memory continuously:

```
import uasyncio
from machine import Pin
import  micropython

ledRed = Pin(6, Pin.OUT)
ledGreen = Pin(7, Pin.OUT)

async def blink(led, period_ms):
    while True:
        led.on()
        await uasyncio.sleep_ms(period_ms)
        led.off()
        await uasyncio.sleep_ms(period_ms)
        
async def info():
    while True:
        print(micropython.mem_info())
        await uasyncio.sleep(2)

async def main():
    uasyncio.create_task(info())
    uasyncio.create_task(blink(ledGreen, 700))
    uasyncio.create_task(blink(ledRed, 400))
    await uasyncio.sleep_ms(20_000) 
    ledRed.off
    ledGreen.off

# Run
uasyncio.run(main())
print('done')
```

**Comment by bixb922** [comment]
Please format code with back ticks, see Welcome post.
mem_free usually decreases until an automatic or manual garbage collection takes place.
Please see https://docs.micropython.org/en/latest/develop/memorymgt.html and https://docs.micropython.org/en/latest/library/gc.html
Does the program crash with an error? Which error? Could you try and tell what happens when mem_free reaches zero?

**Comment by steeley** [comment]
thanks - no idea why my post is not formatted correctly -  I'm  sure I put in the correct back ticks. will edit post.
let me do some more tests and see what happens with the memory.

**Comment by steeley** [comment]
This is what happens.:

Start of program- memory use is:
```
stack: 2516 out of 14336
GC: total: 124032, used: 5280, free: 118752
 No. of 1-blocks: 98, 2-blocks: 17, max blk sz: 18, max free sz: 7419
None
```

After about half and hour the memory gets all used up, and then goes back up to close to the starting values. I guess the GC is doing its thing ok(??).There is no error or crash.

```
stack: 2516 out of 14336
GC: total: 124032, used: 123648, free: 384
 No. of 1-blocks: 7496, 2-blocks: 17, max blk sz: 18, max free sz: 24
None
```

**Comment by peterhinch** [comment, marked as answer]
@steeley The way to check for memory leaks is first to issue `gc.collect()` then to check free RAM, the aim being to detect permanently allocated RAM.

What you are measuring is the normal process whereby blocks of memory are allocated. Subsequently all references to the block of RAM go out of scope. The RAM is still seen as being used until a GC is performed, when the lack of references is detected and the block is reclaimed.

As a general point I have used `uasyncio` extensively in long running applications and have never detected a leak.

**Reply by steeley** [reply]
Ahh, thanks that makes perfect sense. Coincidentally I just found your tutorial stuff for uasyncio - really good!
